ksqldb-engine/src/main/java/io/confluent/ksql/KsqlExecutionContext.java

```diff
@@ -125,6 +129,16 @@ TransientQueryMetadata executeQuery(
       ConfiguredStatement<Query> statement
   );
 
+  PullQueryResult executePullQuery(
```

Comment:
Yes, I thought about this as well. It's a tradeoff between sending larger messages (the physical plan vs the query) and computation. If we do want to make this change, this would be part of a larger refactoring, and not part of this PR. 

The routing already works as you describe, in the sense that server B executes the request locally. If it fails, server A will try another server.

Reply:
Yeah this is totally a meta thing which is not to be addressed in this PR. I left it here just to dump my wild thoughts here for future references.

Regarding the trade-off, I think with some encoding the number of bytes that we need to transmit between servers with the physical plan would not be orders of magnitude larger (my bold claim, of course :P) but the benefit that only one single server makes the compilation decision to achieve consistent behavior could be huge.